pkg/manager/member/failure_recovery.go

```diff
@@ -0,0 +1,337 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package member
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/controller"
+	"github.com/pingcap/tidb-operator/pkg/util"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/klog/v2"
+)
+
+const (
+	// annAutoFailureRecovery in TidbCluster indicates whether auto failure recovery is enabled for the components
+	annAutoFailureRecovery = "kubernetes.io/auto-failure-recovery"
+
+	podPhaseRunning = "Running"
+	podPhaseUnknown = "Unknown"
+
+	// Reason for host down being true
+	hdReasonNodeFailure           = "NodeFailure"
+	hdReasonRODiskFound           = "RODiskFound"
+	hdReasonStoreDownTimeExceeded = "StoreDownTimeExceeded"
+
+	restartToDeleteStoreGap = 10 * time.Minute
```

Comment:
How about increasing thie duration if some new nodes need to be provisioned?

Reply:
In case there is no more capacity available for the new pod in EBS like use case, the pod will be in pending state after the force restart. After this 10 minutes has passed after the force restart action, the operator will go ahead and delete the store since a new pod did not come up because of no capacity. The pending pod will automatically get created when extra capacity is added to the k8s cluster. The 10 minutes here is a hard limit on top of the failover-period for us to delete the store or member (which will then allow faster region balancing to happen in case of Tikv/Tiflash) after k8s node failure is detected.

The rationale for keeping it 10 minutes is that, the pod-eviction-timeout in K8s is 5 minutes, so we have given another 5 more minutes to that for the new pod to come up. We also expect that if it takes more than 10 minutes for new pod to come up, the operator will go ahead and delete the store (so that the region balancing can start to happen in case of Tikv/Tiflash)